halo2_proofs/src/arithmetic.rs

```diff
@@ -115,6 +116,67 @@ fn multiexp_serial<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C], acc: &mut
     }
 }
 
+#[cfg(feature = "no_floating_points")]
+const LN_CEIL_LOOKUP_TABLE: &[usize] = &[
```

Comment:
We can definitely simplify this lookup table:

- It can resolve down to `c = 0`, but that is not necessary because we only use it with `bases.len() >= 32`. So we can drop the first 5 rows of the table (and change `idx.saturating_sub(1)` to `idx + 4`).
- It can resolve up to `c = 45`, but is not necessary for it to be nearly this large, because the `C` type will be at least 64 bytes in memory (it's an affine curve point, so for a 255-bit base field like Pallas or Vesta you'll have two 32-byte field elements for coordinates). If `bases.len() > 86593400423994057` in this table, then a 64-bit memory address space will likely be exhausted; in practice for machines with 16 GiB of RAM, `bases.len >= 134217728` would consume it in entirety. If we set a maximum likely memory usage of 1 TiB for curves with 255-bit base fields, then we can drop the last 22 rows of the table.

The above would reduce the table size from 46 rows to 19. But we should also consider whether we need to use / emulate `f64::ln` here at all, or whether we can instead use `f64::log2`. If that would work, then we could replace all of this (for all targets) with `usize::next_power_of_two` and `usize::trailing_zeros`, using something like:

```rust
    let num_bases = bases.len();
    let c = if num_bases < 4 {
        1
    } else if num_bases < 32 {
        3
    } else {
        num_bases.next_power_of_two().trailing_zeros() - 2 // or whatever adjustment makes sense
    }
```

@ebfull was there any specific reason for using `f64::ln` here?